### ISSUE TITLE
[PATCH] Fix "make check" link errors with LDFLAGS="-Wl,--as-needed"

### ISSUE THREAD

**Issue by hartwork** [mypaint/libmypaint#62]
Hi!

We [ran into](https://bugs.gentoo.org/show_bug.cgi?id=589652) another issue with `-Wl,--as-needed`.
In contrast to issue #58, this issue is not related to GNU gold.

```
# ./autogen.sh && ./configure LDFLAGS="-Wl,--as-needed" --enable-gegl && make clean all check
[..]
Making check in gegl
make[2]: Entering directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
make  check-TESTS
make[3]: Entering directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
make[4]: Entering directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
depbase=`echo test-gegl-surface.o | sed 's|[^/]*$|.deps/&|;s|\.o$||'`;\
gcc -std=gnu99 -DHAVE_CONFIG_H -I. -I../..  -I../.. -I../.. -I../../gegl -I../../tests  -DLIBMYPAINT_TESTING_ABS_TOP_SRCDIR='"/tmp/tmp.0snzncE6pB/libmypaint"' -pthread -I/usr/include/gegl-0.3 -I/usr/include/json-glib-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/babl-0.1 -I../../gegl/ -I.. -D_POSIX_C_SOURCE=200809L -g -O2 -MT test-gegl-surface.o -MD -MP -MF $depbase.Tpo -c -o test-gegl-surface.o test-gegl-surface.c &&\
mv -f $depbase.Tpo $depbase.Po
/bin/sh ../../libtool  --tag=CC   --mode=link gcc -std=gnu99 -DLIBMYPAINT_TESTING_ABS_TOP_SRCDIR='"/tmp/tmp.0snzncE6pB/libmypaint"' -pthread -I/usr/include/gegl-0.3 -I/usr/include/json-glib-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/babl-0.1 -I../../gegl/ -I.. -D_POSIX_C_SOURCE=200809L -g -O2  -Wl,--as-needed -o test-gegl-surface test-gegl-surface.o -lm  -lgegl-0.3 -lgegl-npd-0.3 -lm -Wl,--export-dynamic -lgmodule-2.0 -pthread -ljson-glib-1.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lbabl-0.1 ../../libmypaint.la ../../gegl/libmypaint-gegl.la ../../tests/libmypaint-tests.a -lm 
libtool: link: gcc -std=gnu99 -DLIBMYPAINT_TESTING_ABS_TOP_SRCDIR=\"/tmp/tmp.0snzncE6pB/libmypaint\" -pthread -I/usr/include/gegl-0.3 -I/usr/include/json-glib-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/babl-0.1 -I../../gegl/ -I.. -D_POSIX_C_SOURCE=200809L -g -O2 -Wl,--as-needed -o .libs/test-gegl-surface test-gegl-surface.o -Wl,--export-dynamic -pthread  -lgegl-0.3 -lgegl-npd-0.3 -lgmodule-2.0 -ljson-glib-1.0 -lgio-2.0 -lbabl-0.1 ../../.libs/libmypaint.so -ljson-c -lgobject-2.0 -lglib-2.0 ../../gegl/.libs/libmypaint-gegl.so ../../tests/libmypaint-tests.a -lm -pthread
../../tests/libmypaint-tests.a(mypaint-test-surface.o): In function `test_surface_drawing':
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:58: undefined reference to `mypaint_brush_new'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:61: undefined reference to `mypaint_brush_from_string'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:62: undefined reference to `mypaint_brush_set_base_value'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:77: undefined reference to `mypaint_surface_begin_atomic'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:82: undefined reference to `mypaint_surface_end_atomic'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:95: undefined reference to `mypaint_brush_unref'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-test-surface.c:96: undefined reference to `mypaint_surface_unref'
../../tests/libmypaint-tests.a(mypaint-utils-stroke-player.o): In function `mypaint_utils_stroke_player_iterate':
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-utils-stroke-player.c:142: undefined reference to `mypaint_brush_stroke_to'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-utils-stroke-player.c:148: undefined reference to `mypaint_surface_end_atomic'
/tmp/tmp.0snzncE6pB/libmypaint/tests/mypaint-utils-stroke-player.c:139: undefined reference to `mypaint_surface_begin_atomic'
../../gegl/.libs/libmypaint-gegl.so: undefined reference to `mypaint_tiled_surface_init'
../../gegl/.libs/libmypaint-gegl.so: undefined reference to `mypaint_surface_ref'
../../gegl/.libs/libmypaint-gegl.so: undefined reference to `mypaint_tiled_surface_destroy'
collect2: error: ld returned 1 exit status
make[4]: *** [Makefile:623: test-gegl-surface] Error 1
make[4]: Leaving directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
make[3]: *** [Makefile:839: check-TESTS] Error 2
make[3]: Leaving directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
make[2]: *** [Makefile:909: check-am] Error 2
make[2]: Leaving directory '/tmp/tmp.0snzncE6pB/libmypaint/tests/gegl'
make[1]: *** [Makefile:780: check-recursive] Error 1
[..]
```

Please consider applying the `git-am` patch attached.  I hope you don't mind that way of sharing the patch: It's way easier to me.

[libmypaint-1.3.0_beta1-as-needed.patch](https://github.com/mypaint/libmypaint/files/381520/libmypaint-1.3.0_beta1-as-needed.patch.txt)

Best, Sebastian

**Comment by achadwick** [comment]
We strongly prefer Github PRs. They're way easier to review and apply. Plus CI checks are a thing that way. Your time spent creating them would be appreciated immensely! :smile: 

(Not going to insist on that for now, though. We can always revert.)

**Comment by hartwork** [comment]
Alright, I'll be back to you in a few minutes.

**Comment by achadwick** [comment]
Merged in 94f059dbeaa94991466964d0357e6f6e54bb5ece. Can you re-test please?

**Comment by hartwork** [comment]
Confirmed fixed. Thank you!
